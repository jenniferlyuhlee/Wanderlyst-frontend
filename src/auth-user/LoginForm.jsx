import React, { useState, useContext} from "react";
import { useNavigate, Navigate } from "react-router-dom";
import UserContext from "./UserContext";
import Alert from "../shared/Alert";

/** Login Form component
 * Props: login
 * - When submitted, calls login function prop
 * - Redirects user to user homepage
 */

function LoginForm({login}){
    const initialState = {
        username: "",
        password: ""
    };
    // hooks called unconditionally
    const { currUser } = useContext(UserContext);
    const navigate = useNavigate();
  
    // conditional rendering - redirects logged in users to homepage
       if(currUser){
        return <Navigate to="/" />
    }

    // states: formData, formErrors
    const [formData, setFormData] = useState(initialState);
    const [formErrors, setFormErrors] = useState(null);

    // general callback to update targeted form data
    function handleChange(e){
        const { name, value } = e.target;
        setFormData(data => ({...data, [name]: value}))
    }

    // form submission handler
    async function handleSubmit(e){
        e.preventDefault();
        let result = await login(formData);
        if(result.success){
            navigate("/");
        }
        else{
            setFormErrors(result.err)
        
        }
    }

    return(
        <div className = "Home">
        <div className="container card px-5 auth-form">
            <h1 className = "text-center mb-3">Login</h1>
            <p className = "my-2">Share & stay updated on unforgettable experiences.</p>
            <form onSubmit = {handleSubmit} className = "row">
                <div className = "form-floating mb-3 p-0">
                    <input id = "username"
                        className = "form-control"
                        name = "username"
                        placeholder = "username"
                        value = {formData.username}
                        onChange = {handleChange}
                        required
                    />
                    <label htmlFor = "username">Username</label>
                </div>
                <div className = "form-floating mb-3 p-0">
                    <input id = "password"
                        className = "form-control"
                        name = "password"
                        type = "password"
                        placeholder=""
                        value = {formData.password}
                        onChange = {handleChange}
                        required
                    />
                    <label htmlFor = "password">Password</label>
                </div>
                {formErrors ?
                    <Alert type = "danger" messages = {formErrors}/>
                    :
                    null
                }
                <button className = "btn rounded-pill btn-lg btn-primary">
                    Login
                </button>
            </form>
        </div>
        </div>
    )
}

export default LoginForm;