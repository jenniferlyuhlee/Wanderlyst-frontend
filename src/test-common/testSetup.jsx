import React, { useState } from "react";
import UserContext from "../auth-user/UserContext";
import WanderlystApi from "../utils/api";
import '@testing-library/jest-dom';

const demoUser = {
  username: "testadmin", 
  firstName: "testfirst",
  lastName: "testlast",
  location: "testloc",
  bio: "testbio",
  profilePic: null,
  isAdmin: true,
};

const demoTags = [
  {id: 1, name: 'Foodie', description: 'Savor the world one bite at a time. Dive into a journey packed with delicious local cuisines, hidden culinary gems, and must-try dishes that will leave your taste buds tingling.'},
  {id: 2, name: 'Explorer', description: 'Step off the beaten path and uncover the unknown. From ancient ruins to vibrant cityscapes, let your curiosity guide you through unforgettable experiences.'}
]

const UserProvider = ({ children, 
                        currUser = demoUser, 
                        setCurrUser, 
                        tags = demoTags,
                        logout,
                        hasLikedItin,
                        toggleLike }) => {
  const [user, setUser] = useState(currUser);

  return (
    <UserContext.Provider value={{ currUser: user, 
                                  setCurrUser: setUser, 
                                  tags, logout, 
                                  hasLikedItin, toggleLike}}>
      {children}
    </UserContext.Provider>
  );
};

export { UserProvider };