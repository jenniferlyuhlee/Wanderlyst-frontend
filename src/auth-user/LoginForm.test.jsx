import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import { UserProvider } from "../test-common/testSetup";
import { MemoryRouter } from "react-router-dom";
import LoginForm from "./LoginForm";
import { beforeEach, afterEach, expect, describe, test, vi } from "vitest";

const mockLogin = vi.fn().mockResolvedValue({success: true});

describe("LoginForm Component", () => {
    beforeEach(()=> {
        mockLogin.mockClear();
    });

    test("renders without crashing: anon", () => {
        render(
        <MemoryRouter>
            <UserProvider currUser = {null}>
                <LoginForm login={mockLogin} />
            </UserProvider>
        </MemoryRouter>)
    });
            
    
    test("matches snapshot", () => {
        const {asFragment} = render(
                            <MemoryRouter>
                                <UserProvider currUser = {null}>
                                    <LoginForm login={mockLogin}/>
                                </UserProvider>
                            </MemoryRouter>)
        expect(asFragment()).toMatchSnapshot();
    });

    test("calls passed function with inputted values", async () => {
        const { getByRole, getByLabelText } = render(
            <MemoryRouter>
                <UserProvider currUser = {null}>
                    <LoginForm login={mockLogin} />
                </UserProvider>
            </MemoryRouter>);
       
        // capture fields and login button
        const username = getByLabelText("Username");
        const password = getByLabelText("Password");
        const btn = getByRole("button");

        // fill in fields and submit
        fireEvent.change(username, {target: {value: 'testuser'}});
        fireEvent.change(password, {target: {value: 'password'}});
        fireEvent.click(btn);
        
        await waitFor(() => expect(mockLogin).toHaveBeenCalled());
        await waitFor(() => expect(mockLogin).toHaveBeenCalledWith({
          username: 'testuser',
          password: 'password'
        }));

    });
})
