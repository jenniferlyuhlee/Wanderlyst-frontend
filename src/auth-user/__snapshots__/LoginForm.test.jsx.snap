// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginForm Component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Home"
  >
    <div
      class="container card px-5 auth-form"
    >
      <h1
        class="text-center mb-3"
      >
        Login
      </h1>
      <p
        class="my-2"
      >
        Share & stay updated on unforgettable experiences.
      </p>
      <form
        class="row"
      >
        <div
          class="form-floating mb-3 p-0"
        >
          <input
            class="form-control"
            id="username"
            name="username"
            placeholder="username"
            required=""
            value=""
          />
          <label
            for="username"
          >
            Username
          </label>
        </div>
        <div
          class="form-floating mb-3 p-0"
        >
          <input
            class="form-control"
            id="password"
            name="password"
            placeholder=""
            required=""
            type="password"
            value=""
          />
          <label
            for="password"
          >
            Password
          </label>
        </div>
        <button
          class="btn rounded-pill btn-lg btn-primary"
        >
          Login
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
